### ISSUE TITLE
[intake] Complete ProjectOdyssey issue #3181: implement the full backward

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#5511]
## Intake

**Idea:** Complete ProjectOdyssey issue #3181: implement the full backward pass for the ResNet-18 CIFAR-10 training script and validate that it trains an epoch

**Context:** Target repo HomericIntelligence/ProjectOdyssey. Issue #3181 already carries state:plan-go with an approved plan: full backward pass through all residual blocks, batch_norm2d_backward in shared/core/normalization.mojo, activation caching in the forward pass, 84 SGD-momentum velocity tensors, parameter updates, and training-convergence validation on CIFAR-10. Describe the work as at most 3 dependency-ordered tasks for pipeline.task-agent; the final PR must include 'Closes #3181'; validation must produce a training log for one epoch (bounded CIFAR-10 subset acceptable if a full epoch exceeds 60 min on CPU).

Interview transcript and researched brief follow as comments.

**Comment by mvillmow** [comment]
# Implementation Plan

I now have everything needed. The plan is a verification-and-gap-closure plan grounded in the actual merged state. Writing it.

# Implementation Plan

## Objective

Issue #5511 asks to "complete ProjectOdyssey issue #3181: implement the full backward pass for the ResNet-18 CIFAR-10 training script and validate that it trains an epoch." Investigation shows **#3181 is already CLOSED and all its deliverables are merged to `main`**: the full manual backward pass, `batch_norm2d_backward` usage, activation caching, the named velocity struct, per-parameter SGD-momentum updates, and a one-epoch CIFAR-10 validation. The single genuine open gap is the **ADR-014 evidence gate**: `examples/resnet18_cifar10/validation/epoch1.log` is a *hand-committed metric artifact* with no machine-checked reproduction path — the summary itself defers that to unbuilt CI smoke **#5551**. This plan (a) proves the merged backward implementation compiles and its backward tests pass on this checkout, then (b) closes the evidence gap by adding a CI training-entrypoint smoke that machine-checks the entrypoint emits finite, monotonically-decreasing loss on a bounded subset — without hand-authoring or committing any metric.

## Approach

**Decision 1 — This is verification + gap-closure, not implementation.** The backward pass is already present and wired.
- `examples/resnet18_cifar10/train.mojo:126` `backward_identity_block`, `:201` `backward_projection_block`, `:292` `train_step` (full loss→logits→FC→GAP→stage4..stage1→stem backward), `:1056` `train_epoch`, `:1130` `main` (real `CIFAR10Dataset`).
- Merged via `git log -- examples/resnet18_cifar10/train.mojo`: #5538 (backward + SGD), #5546 (convergence validation), #5548 (real CIFAR-10 + one-epoch evidence).
- Issue #3181: `gh issue view 3181 --json state` → `CLOSED`. Issue #3076 (data-loading blocker cited in the old plan) → `CLOSED`.
- Re-implementing would violate KISS/YAGNI and the "look at the target before overwriting" rule. The plan does **not** touch the backward math.

**Decision 2 — Backward tests already run in CI; the plan re-verifies, not re-authors them.** `examples/resnet18_cifar10/test_backward.mojo` (3 tests: runs-without-error, every-stage-nonzero-gradient, loss-decreases) is executed by `just test-example-backward` → `comprehensive-tests.yml:162` (`example-backward-tests` job, added by the latest commit `4381bddb`). `test_forward_cache_velocities.mojo` covers the 82-field velocity struct and cache population. No new backward unit test is needed.

**Decision 3 — Close the ADR-014 gap (#5551), the only real open work.** `validation/epoch1-summary.md:34-35` explicitly states the committed log is an *artifact, not a gate*, and that reproduction is deferred to "the planned CI training-smoke (ProjectOdyssey #5551)". `gh issue view 5551` → **OPEN**. `grep -rn "training-smoke\|smoke.*train" .github/workflows/ justfile` → no such job exists yet. Per the RUNNABLE-EVIDENCE RULE, I add a CI job that runs the *training entrypoint itself* on a small bounded subset and machine-asserts the run's own emitted output is finite and decreasing. The asserting harness reads only verbatim stdout from a run the plan does not author — no metric is hand-written or committed.

**Decision 4 — Subset the smoke to fit the CI timeout, following the existing job pattern.** The full epoch took ~17.8h on CPU (`epoch1-summary.md`); a full run cannot gate a PR. The smoke trains on a tiny fixed subset (a few dozen batches) so it completes inside a `timeout-minutes` budget matching the sibling `example-backward-tests` job (20 min). It asserts *direction and finiteness* (loss finite, last-logged < first-logged), not a specific accuracy number — the honest, reproducible claim. This is exactly the guarantee `test_loss_decreases_over_steps` already makes at unit scale, promoted to the real entrypoint at CI scale.

## Files to Create

_None required for the core deliverable._ The training-smoke is added as a justfile recipe + CI job (modifications below). A new bounded-subset smoke test file is created **only if** `train.mojo:main` cannot be invoked with a small-subset flag (determined in step 3); if a new file is needed:

### `examples/resnet18_cifar10/test_train_smoke.mojo` (conditional — only if `main` lacks a subset path)
A `def main() raises` harness that: builds `ResNet18`, synthesizes or slices a small (~32-sample, 2-class) CIFAR-10-shaped batch, runs `train_epoch` for N≈20 steps recording `loss_history`, and asserts (via the `assert_true` pattern already in `test_backward.mojo:14`) that `not isnan/isinf(loss_history[-1])` and `loss_history[-1] < loss_history[0]`. This is a machine check of the entrypoint's own output, committing no metric. It reuses the separable-batch builder pattern from `test_backward.mojo:95 _build_separable_batch`.

## Files to Modify

### `justfile`
Add a `test-resnet-train-smoke` recipe mirroring the private-inner pattern of `_test-example-backward-inner` at `justfile:1299`. Insert after the `test-example-backward` block (~`justfile:1330`):

```makefile
# Run the ResNet-18 training-entrypoint smoke (ADR-014 metric gate, #5551).
# Executes the real training path on a bounded subset and machine-checks the
# run's own emitted loss is finite and decreasing. Commits no metric.
test-resnet-train-smoke:
    @just _run "just _test-resnet-train-smoke-inner"

[private]
_test-resnet-train-smoke-inner:
    #!/usr/bin/env bash
    set -euo pipefail
    REPO_ROOT="$(pwd)"
    f="examples/resnet18_cifar10/test_train_smoke.mojo"
    [ -f "$f" ] || { echo "MISSING (expected): $f"; exit 1; }
    pixi run mojo run --Werror \
        -I "$REPO_ROOT/src" -I "$REPO_ROOT" \
        -I "$REPO_ROOT/examples/resnet18_cifar10" -Xlinker -lm "$f"
```

(If step 3 shows `main` accepts a subset/`--max-batches` flag, the recipe instead runs the compiled `train.mojo` entrypoint with that flag and pipes stdout to a `grep`/`awk` assertion on the `Loss:` lines — no new `.mojo` file.)

### `.github/workflows/comprehensive-tests.yml`
Add a `resnet-train-smoke` job cloned from the `example-backward-tests` job at `comprehensive-tests.yml:162` (same `needs`, `setup-container`, `setup-just`, `timeout-minutes: 20`), changing only the name and run step:

```yaml
  resnet-train-smoke:
    needs: [mojo-compilation]
    runs-on: ubuntu-latest
    timeout-minutes: 20
    name: "ResNet-18 Training Smoke (ADR-014)"
    steps:
      - name: Checkout code
        uses: actions/checkout@9c091bb21b7c1c1d1991bb908d89e4e9dddfe3e0  # v7.0.0
      - name: Set up container
        uses: ./.github/actions/setup-container
      - name: Install Just
        uses: extractions/setup-just@53165ef7e734c5c07cb06b3c8e7b647c5aa16db3  # v4.0.0
      - name: Run ResNet-18 training smoke
        run: just test-resnet-train-smoke
```

### `examples/resnet18_cifar10/validation/epoch1-summary.md`
Update lines `34-35` once the smoke exists: change "the planned CI training-smoke (ProjectOdyssey #5551) **will** execute" → "the CI training-smoke (`resnet-train-smoke`, #5551) executes on every PR to machine-check that a real run emits finite, decreasing loss." This closes the ADR-014 loop the summary itself opened.

## Implementation Order

1. **Verify merged state compiles** — build the shared package and the example, confirming the already-merged backward code is sound on this checkout (no edits): `just build`.
2. **Run the existing backward + cache tests** — confirm the merged implementation passes as-is before adding anything: `just test-example-backward` and the forward-cache test.
3. **Probe the entrypoint's subset capability** — read `parse_training_args_with_defaults` usage in `train.mojo:main` (`:1130`) to decide whether a `--max-batches`/subset flag exists or a small `test_train_smoke.mojo` harness is needed. This picks between the two `justfile` recipe variants above.
4. **Add the smoke** — create the recipe (+ conditional `test_train_smoke.mojo`) and run it locally in-container to confirm it completes inside 20 min and its assertions pass.
5. **Wire CI** — add the `resnet-train-smoke` job to `comprehensive-tests.yml`.
6. **Update the ADR-014 note** in `epoch1-summary.md` to reference the now-real gate.
7. **PR** — one PR, `Closes #3181` (as #5511 mandates) and `Closes #5551` (the gap actually closed). Enable auto-merge; wait for CI green.

## Verification

Run in-container (host GLIBC may be < 2.32 per CLAUDE.md); use `just shell` then the recipes.

```bash
# Criterion "full backward pass compiles" (#3181 deliverable, already merged):
just build                                                          # shared pkg + example compile clean

# Criterion "backward pass through all residual blocks + BN backward + activation
# caching + 84/82 velocity tensors + parameter updates" (#3181 deliverables):
just test-example-backward                                          # test_backward.mojo: runs-without-error,
                                                                    #   every-stage-nonzero-gradient, loss-decreases
pixi run mojo run --Werror -I src -I . -I examples/resnet18_cifar10 -Xlinker -lm \
  examples/resnet18_cifar10/test_forward_cache_velocities.mojo      # velocity-struct fields + cache population

# Criterion "validate that it trains an epoch" (#3181 + #5511), machine-checked (ADR-014):
just test-resnet-train-smoke                                        # entrypoint emits finite, decreasing loss on
                                                                    #   a bounded subset — asserts direction, not a number
```

**ADR-014 gating note (RUNNABLE-EVIDENCE RULE):** The *full-epoch* metric numbers (60.02% top-1, epoch loss 1.378) in `epoch1.log` are NOT a deliverable this task re-produces or re-commits — a genuine epoch exceeds the session/CI budget (~17.8h). This plan's measured evidence is confined to the **bounded-subset smoke's own emitted output**, asserted by a harness the plan does not hand-feed. The existing committed `epoch1.log` is left as the historical artifact it already is; no new log or metric file is authored or committed.

## Skills Used

- **`/mnemosyne:advise` (team knowledge base):** The Prior Learnings block returned `{"skills": []}` — empty marketplace, no registered skills matched this backward-pass intake, so none were applied.
- Local memory feedback applied to the plan's PR/CI posture: **no-admin-merge** (auto-merge + wait for CI, never `--admin`), **fix-don't-disable** (the ADR-014 gate fixes the evidence gap rather than suppressing it), **validate-before-claiming** (steps 1–2 verify the merged code compiles/passes on this checkout before the plan asserts #3181 is complete).

## Changes from Review

_N/A — initial plan._ (Note: a prior *plan draft* is prepended in context, but no `## 🔍 Plan Review` block is present, so this is the first reviewable plan. This plan supersedes that draft because its premise — a placeholder `compute_batch_gradients()` needing implementation — is stale: the backward pass, velocity struct, SGD updates, and one-epoch validation are already merged and #3181 is closed. The open work is the ADR-014 CI evidence gate (#5551), which this plan addresses.)

**Comment by mvillmow** [comment]
Closing as duplicate: triplicate intake (identical intake UUID e088d828, filed 3× within 5s on 2026-07-02) targeting ProjectOdyssey #3181, which is already CLOSED ("Full backward pass for ResNet-18 training script"). ResNet-18 is additionally ported to the autograd path in #5561 / PR #5593. No remaining work.
